adam/perception/high_level_semantics_situation_to_developmental_primitive_perception.py

```diff
@@ -136,17 +146,117 @@ def do(self) -> PerceptualRepresentation[DevelopmentalPrimitivePerceptionFrame]:
+        action_roles_to_fillers = cast(
+            ImmutableSetMultiDict[OntologyNode, SituationObject],
+            situation_action.argument_roles_to_fillers,
+        )
+
+        # TODO: Handle finding manipulator
+
+        before_relations = self._get_relations_from_condition(
+            conditions=action_description.preconditions,
+            action_description=action_description,
+            action_roles_to_fillers=action_roles_to_fillers,
+        )
+        after_relations = self._get_relations_from_condition(
+            conditions=action_description.postconditions,
+            action_description=action_description,
+            action_roles_to_fillers=action_roles_to_fillers,
         )
 
+        return before_relations, after_relations
+
+    def _get_relations_from_condition(
+        self,
+        conditions: ImmutableSet[SituationRelation],
+        action_description: ActionDescription,
+        action_roles_to_fillers: ImmutableSetMultiDict[OntologyNode, SituationObject],
+    ):
+        # TODO: Handle multiple frames (i.e. "mom moved" vs "mom moved the box")
+        entities_to_roles = action_description.frames[0].entities_to_roles
+        for condition in conditions:  # each one is a SituationRelation
+            relation_type, first_slot, second_slot, negated = (
+                condition.relation_type,
+                condition.first_slot,
+                condition.second_slot,
+                condition.negated,
+            )
+
+            # TODO: Handle multiple semantic roles
+            if (
+                len(entities_to_roles[first_slot]) > 1
+                or len(action_roles_to_fillers[list(entities_to_roles[first_slot])[0]])
+                > 1
+            ):
+                raise RuntimeError("Can not handle multiple semantic roles")
+            situation_object_1 = list(
+                action_roles_to_fillers[list(entities_to_roles[first_slot])[0]]
+            )[
+                0
+            ]  # e.g. mom
+
+            if (
+                len(entities_to_roles[second_slot]) > 1
+                or len(action_roles_to_fillers[list(entities_to_roles[second_slot])[0]])
+                > 1
+            ):
+                raise RuntimeError("Can not handle multiple semantic roles")
+            situation_object_2 = list(
+                action_roles_to_fillers[list(entities_to_roles[second_slot])[0]]
+            )[
+                0
+            ]  # e.g. ball
+
+            perception_1 = self._objects_to_perceptions[situation_object_1]
+            perception_2 = self._objects_to_perceptions[situation_object_2]
+            relation_perception = RelationPerception(
+                relation_type=relation_type, arg1=perception_1, arg2=perception_2
+            )
+
+            # TODO: Implement negation
+            if not negated:
+                self._relation_perceptions.append(relation_perception)
+
+        return self._relation_perceptions.copy()
```

Comment:
It might be cleaner to not store `_relation_perceptions` on the object and just computer and return them here. For the second frame, you could add an optional argument for "already known relations" or something so you can pass these in.